apm-lambda-extension/logsapi/subscribe.go

```diff
@@ -0,0 +1,111 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package logsapi
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+var listenerHost = "sandbox"
+var logsAPIServer *http.Server
+var logsAPIListener net.Listener
+
+type LogEvent struct {
+	Time      time.Time       `json:"time"`
+	Type      string          `json:"type"`
+	RawRecord json.RawMessage `json:"record"`
+	Record    LogEventRecord
+}
+
+type LogEventRecord struct {
+	RequestId string `json:"requestId"`
+	Status    string `json:"status"`
+}
+
+// Subscribes to the Logs API
+func subscribe(extensionID string, eventTypes []EventType) error {
+	extensions_api_address, ok := os.LookupEnv("AWS_LAMBDA_RUNTIME_API")
+	if !ok {
+		return errors.New("AWS_LAMBDA_RUNTIME_API is not set")
+	}
+
+	logsAPIBaseUrl := fmt.Sprintf("http://%s", extensions_api_address)
+	logsAPIClient, err := NewClient(logsAPIBaseUrl)
+	if err != nil {
+		return err
+	}
+
+	_, port, _ := net.SplitHostPort(logsAPIListener.Addr().String())
+	_, err = logsAPIClient.Subscribe(eventTypes, URI("http://"+listenerHost+":"+port), extensionID)
+	return err
+}
+
+// Starts the HTTP server listening for log events and subscribes to the Logs API
+func Subscribe(ctx context.Context, extensionID string, eventTypes []EventType, out chan LogEvent) (err error) {
+	if checkAwsSamLocal() {
+		return errors.New("Detected sam local environment")
+	}
+	err = startHttpServer(out)
+	if err != nil {
+		return
+	}
+
+	err = subscribe(extensionID, eventTypes)
+	if err != nil {
+		return
+	}
+	return nil
+}
+
+func startHttpServer(out chan LogEvent) (err error) {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/", handleLogEventsRequest(out))
+
+	logsAPIServer = &http.Server{
+		Handler: mux,
+	}
+
+	logsAPIListener, err = net.Listen("tcp", listenerHost+":0")
+	if err != nil {
+		return
```

Comment:
It's typically best practice to `return err` and not define a variable for the return types in the function signature, ie.

`startHTTPServer(out chan LogEvent) error`

Reply:
I thought setting the error to the variable `err` in the signature was an idiomatic way to allow you to use shorthand and just `return` in the function, as it is done in line 94. But maybe it's just a question of personal style?

Reply:
made the change anyway :)